### DISCUSSION TITLE
Definitie van Component

### DISCUSSION THREAD

**Discussion by RonaldvCortenberghe** [OpenCatalogi/.github#104] [Q&A]
Hierbij een eerste aanzet (uw kritische bijdrage is gewenst). 

OpenCatalogi is een weergave van Componenten verdeeld over de 5 lagen zoals gedefinieerd door VNG in het Gegevenslandschap (https://www.gemmaonline.nl/index.php/Gegevenslandschap en https://componentencatalogus.commonground.nl/5-lagen-model).

### COMPONENT
Voor OpenCatalogi hanteren we de archimate definitie _met de aanscherpingen_ die vanuit de VNG worden meegegeven in hoofdstuk 2.1 van https://www.gemmaonline.nl/images/gemmaonline/0/09/GEMMA_Gegevenslandschap_-_Informatiearchitectuurprincipes_v1_0.pdf.

De archimate definitie zoals vertaald door VNG: 
Een modulair, zelfstandig inzetbaar, herbruikbaar en vervangbaar deel van een systeem, dat zijn functionaliteit aanbiedt via goed gedefinieerde interfaces. Applicatiecomponenten stellen functionaliteit beschikbaar, die gebruikt wordt om de applicatiediensten mee te leveren. 

De aanscherpingen vanuit de VNG: 
We werken met componenten die **afgebakende functionaliteit** kennen, **zoveel mogelijk ontkoppeld** zijn en met 
elkaar communiceren via **gestandaardiseerde interfaces**. Ingevuld met (voor zover van toepassing voor OpenCatalogi):

Gangbare service-oriëntatie
- Autonomie - Componenten zijn zelfstandig inzetbaar, schaalbaar en vervangbaar.
- Losse koppeling - Componenten zijn onderling onafhankelijk en zelfstandig door te ontwikkelen en gebruiken. 
- Herbruikbaar - Componenten zijn vaker in te zetten door ze in verschillende omgevingen te gebruiken of door een component vaker te instantiëren binnen één omgeving.
- Samenstelbaar - Componenten kunnen worden gecombineerd tot nieuwe componenten. 
- Abstractie - Afnemers hoeven alleen de functionaliteit van een component te kennen en niet de interne werking ervan. 
- Service-contract - Een component kent duidelijke voorwaarden en leveringsafspraken over de kwaliteit en het volume van mogelijk gebruik door afnemers.
- Vindbaar - Services zijn voorzien van metadata en zijn goed vindbaar.
- Toestandsloos - Componenten die diensten leveren bewaren geen statusinformatie (maar laten dit over aan afnemers of gespecialiseerde componenten).

Componenten conformeren zich aan het vijf-lagenmodel

- Componenten beperken zich tot functionaliteit binnen de laag waartoe ze behoren 
- Componenten voor interactie- en procesfunctionaliteit vragen gegevens op bij de bron via gestandaardiseerde interfaces.
- Componenten bewerken geen brongegevens anders dan bij de bron 

_Verschijningsvorm per laag_
De verschijningsvorm van een component is afhankelijk van de laag. In navolgende figuur is dit weergegeven:

![image](https://user-images.githubusercontent.com/38490649/177576024-c883e62b-a917-4368-ab0c-7647326c1ac5.png)

**Comment by EdwinCoster** [comment]
Uit de informatiearchitectuur[principes is bij 'Componenten beperken zich tot functionaliteit binnen de laag waartoe ze behoren' ook opgenomen 'Als er goede redenen voor zijn, mag een component conform het pas-toe-of-leg-uit principe ook functionaliteit uit de onderliggende laag realiseren (bijvoorbeeld een front-end-applicatie die ook procesfunctionaliteit bevat omdat daarvoor geen bruikbare component beschikbaar is)'.
Ik ga ervan uit dat dit bewust is.

Het is daarbij natuurlijk mogelijk om af te wijken van een principe.
